realkd/weight_update_methods.py

```diff
@@ -0,0 +1,98 @@
+from typing import Any, Callable, Dict, Optional, Union
+import numpy as np
+from numpy import floating
+from numpy.typing import NDArray
+
+from realkd.rules import AdditiveRuleEnsemble, Rule
+from realkd.correction_methods import get_correction_method
+
+WeightUpdateMethod = Callable[[AdditiveRuleEnsemble, Any, Optional[Union[str, Callable]]], NDArray[floating]]
+
+def get_gradient(g, y, q_mat, weights: NDArray[floating], reg):
+    def gradient(weight):
+        all_weights = np.append(weights, weight)
+        grad_vec = g(y, q_mat.dot(all_weights))
+        return np.array([(q_mat.T.dot(grad_vec) + reg * all_weights)[-1]])
+
+    return gradient
+
+def get_risk(loss, y, q_mat, weights: NDArray[floating], reg):
+    def sum_loss(weight):
+        all_weights = np.append(weights, weight)
+        return sum(loss(y, q_mat.dot(all_weights))) + reg * sum(all_weights * all_weights) / 2
+
+    return sum_loss
+
+def get_hessian(h, y, q_mat, reg):
+    def hessian(weights):
+        h_vec = h(y, q_mat.dot(weights))
+        return q_mat.T.dot(np.diag(h_vec)).dot(q_mat) + np.diag([reg] * len(weights))
+
+    return hessian
+
+def fully_corrective(rules, loss, weight_update_method_params, data, target, reg):
+    '''
+        FullyCorrective updates all weights
+    '''
+    if weight_update_method_params is None:
+        weight_update_method_params = {'correction_method': 'Newton-CG'}
+
+    w = np.array([3.5 if r.y > 100 and loss == 'poisson' else r.y for r in rules])
+
+    q_mat = np.column_stack([rules[i].q(data) + np.zeros(len(data)) for i in range(len(rules))])
+
+    gradient = get_gradient(loss.g, target, q_mat, np.array([]), reg)
+    sum_loss = get_risk(loss, target, q_mat, np.array([]), reg)
+    hessian = get_hessian(loss.h, target, q_mat, reg)
+
+    w = get_correction_method(weight_update_method_params['correction_method'])(w, gradient, sum_loss, hessian)
+
+    return w
+
+def line_search(rules, loss, weight_update_method_params, data, target, reg):
+    '''
+        Line search only updates the most recent weight
+    '''
+    if weight_update_method_params is None:
+        weight_update_method_params = {'correction_method': 'Newton-CG'}
+
+    if weight_update_method_params['correction_method'] != 'GD':
+        # TODO: Not sure if this is correct
+        raise NotImplementedError()
```

Comment:
Is this correct? This matches Fan's implementation, but since it's easy do we want to support other methods?